### PR TITLE
Add qwen2moe

### PR DESCRIPTION
**Pull request by simonJJJ**
This PR adds the support of codes for the coming Qwen2 MoE models [hf](https://github.com/huggingface/transformers/pull/29377).
I changed several macro values to support the **60** experts setting. @ggerganov

### CODE REVIEW

File: ggml-backend.c
Comment (ggerganov) [inline ggml-backend.c:1007]: What was the reason to increase this?

File: ggml.h
Comment (ggerganov) [inline ggml.h:230]: Hm, we are not well prepared for 60-expert MoE models

This change will more than double the size of `ggml_tensor` and I still don't know how to support many source buffers in Metal:

https://github.com/ggerganov/llama.cpp/blob/131b0584096ee9df4d07cb28759dfea6efe6475f/ggml-metal.m#L1750-L1759

What are our options here? cc @slaren

File: ggml-backend.c
Comment (simonJJJ) [reply ggml-backend.c:1007]: It's for the graph hash size.

Comment (simonJJJ) [reply ggml-backend.c:1007]: I just increase and test step-by-step from 16->32->64->128->256, and 256 works.

File: ggml.h
Comment (simonJJJ) [reply ggml.h:230]: The size of each expert is very small. The ffn_moe_intermediate size is divided by 4.

Comment (slaren) [reply ggml.h:230]: We could store all the experts in the same tensor, I think it would be easy to adapt the implementations.

Comment (ggerganov) [reply ggml.h:230]: Should we merge this PR first and then try to resolve this?

File: ggml-backend.c
Comment (slaren) [reply ggml-backend.c:1007]: What is the problem that you are trying to solve with this change? This seems like a workaround for a different issue. What would be the correct fix? This constant will be removed in the short term in favor of dynamically allocating the array of splits, so whatever issue this is trying to workaround, we need to deal with that directly.

File: ggml.h
Comment (slaren) [reply ggml.h:230]: I am not sure if we need to merge this right now. As it is, this will only work with the CPU backend, the sort operators in the GPU backends also require the number of experts to be a power of two, so assuming that it is a large model, this implementation may not be very useful. When we have a model to test with, we can improve the implementation.

Comment (sorasoras) [reply ggml.h:230]: I remember qwen2 has 1.8BX16. that make sense

File: ggml-backend.c
Comment (simonJJJ) [reply ggml-backend.c:1007]: The problem is when calling ggml_gallocr_reserve_n in ggml_backend_sched_reserve, the hash value conflicts.

File: ggml.h
Comment (simonJJJ) [reply ggml.h:230]: @slaren do you have some implementations to deal with this **extra large number** experts issue?

**Comment by simonJJJ** [comment]
@ggerganov @slaren is there any update?

**Comment by sorasoras** [comment]
@simonJJJ @ggerganov  
https://huggingface.co/Qwen/Qwen1.5-MoE-A2.7B  is released.
so any update?
we can start testing it.

**Comment by andy-zhangtao** [comment]
This PR looks awesome, can't wait for it to get merged into the main branch. Could you quickly take a look at the failed test case and patch it up?

**Comment by jpmottin** [comment]
I just tested the "Add qwen2moe #6074" PR... 

**CONVERT**

I try to convert the HF model : 
` python3 convert.py ../../Models/Qwen1.5-MoE-A2.7B --outfile ../../models/Qwen1.5-MoE-A2.7B_1.gguf --vocab-type bpe --p
ad-vocab`

I got this error : 

```
Traceback (most recent call last):
  File "convert.py", line 1486, in <module>
    main()
  File "convert.py", line 1472, in main
    model   = convert_model_names(model, params, args.skip_unknown)
  File "convert.py", line 1217, in convert_model_names
    raise Exception("Unexpected tensor name: [tensor_name]. Use --skip-unknown to ignore it")
Exception: Unexpected tensor name: [tensor_name]. Use --skip-unknown to ignore it
```

To make it work, I have added : `--skip-unknown`

**USAGE**

I then had an other issue, number of experts not correct (0), so I modified/added properties in the config.json of the model.
```
  "num_local_experts":60,
  "num_local_experts_per_tok":60,
  "num_experts_per_tok": 4,
  "num_experts": 60,
  "moe": {
    "num_experts_per_tok": 4,
    "num_experts": 60
  },
```

**Now I got this ERROR**
```
llm_load_tensors: ggml ctx size =    7.10 MiB
llama_model_load: error loading model: create_tensor: tensor 'blk.0.ffn_gate.0.weight' has wrong shape; expected  2048,  5632, got  2048,  1408,     1,     1
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model '../../Models/Qwen1.5-MoE-A2.7B_1.gguf'
{"function":"load_model","level":"ERR","line":678,"model":"../../Models/Qwen1.5-MoE-A2.7B_1.gguf","msg":"unable to load model","tid":"0x1dc4cdc40","timestamp":1711809559}
```

So far, I don't know yet how to fix this 😞 I hope to help others with this comment

**Comment by foldl** [comment]
> We have to resolve the issues outlined earlier in the discussion before we can merge this. Sorry for the delay, but without a proper CUDA and Metal support it does not make sense to merge. Now that there is a model available, it will be easier to complete

I hope below update could be merged to `ggml`.  

`#define GGML_MAX_SRC            62` 

I have added this model to ChatLLM.cpp. It works for CPU. Maybe add warning when `GGML_MAX_SRC` is larger when building for CUDA or Metal at present? After CUDA and Metal are supported, the warning can be removed.

**Comment by slaren** [comment]
> I hope below update could be merged to `ggml`.
> 
> `#define GGML_MAX_SRC 62`

It can't. Instead, we are moving all the experts to the same tensor. The work is being done in #6387.

**Comment by ggerganov** [comment]
@simonJJJ With #6387 now merged, this PR should be adapted and we can merge. Let us know if you would be able to take a look

**Comment by simonJJJ** [comment]
@ggerganov just updated! hope to merge asap.

File: ggml-backend.c
Comment (ggerganov) [inline ggml-backend.c:1026]: minor : clean-up these whitespaces

File: ggml.h
Comment (ggerganov) [inline ggml.h:231]: This should no longer be necessary

File: llama.cpp
Comment (ggerganov) [inline llama.cpp:765]: Align vertically

File: llama.cpp
Comment (phymbert) [inline llama.cpp:8609]: There is now `build_moe_ffn` helper function for mixtral, grok, and dbrx

File: llama.cpp
Comment (phymbert) [inline llama.cpp:3919]: We introduced moe model type, see e_model::MODEL_8X7B, for example

File: llama.cpp
Comment (simonJJJ) [reply llama.cpp:8609]: the arch is not the same. qwen2moe has shared experts in each block.

**Review by ggerganov — APPROVED** [review]

File: llama.cpp
Comment (slaren) [inline llama.cpp:8650]: Is this meant to be equivalent to this?
https://github.com/bozheng-hit/transformers/blob/d9dc993fdd6e6b0a61fe68ccbe838e00c73b9f80/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L860-L863

File: llama.cpp
Comment (slaren) [inline llama.cpp:8544]: This is already done in `llm_build_kv`, it's not necessary to do it here.

File: llama.cpp
Comment (slaren) [reply llama.cpp:8610]: The moe ffn still looks the same to me, and could be implemented using `build_moe_ffn`.

File: llama.cpp
Comment (simonJJJ) [reply llama.cpp:8650]: "logits" ∈(−∞,∞) and "probs" ∈[0,1] look good to me. I can not find better options here.

File: llama.cpp
Comment (simonJJJ) [reply llama.cpp:8610]: another diff is the "weights" is not normalized in qwen2moe.

**Comment by ggerganov** [comment]
I changed the "shared_exp" suffix to the shorter "shexp": [f88e684](https://github.com/ggerganov/llama.cpp/pull/6074/commits/f88e6844a49cb4a6e80793604f412bd815be6ce5)

Seems more inline with the naming convention. This requires to reconvert the models.
@simonJJJ Any objections to this change?

**Comment by simonJJJ** [comment]
> I changed the "shared_exp" suffix to the shorter "shexp": [f88e684](https://github.com/ggerganov/llama.cpp/pull/6074/commits/f88e6844a49cb4a6e80793604f412bd815be6ce5)
> 
> Seems more inline with the naming convention. This requires to reconvert the models. @simonJJJ Any objections to this change?

no

**Comment by simonJJJ** [comment]
@ggerganov is there any check hinders the merging?

**Comment by ggerganov** [comment]
> @ggerganov is there any check hinders the merging?

Should be ready soon - slaren will merge when done with the review

File: llama.cpp
Comment (slaren) [inline llama.cpp:1754]: This also needs an entry in `llama_model_type_name` for its string representation.

**Review by slaren — APPROVED** [review]

**Pull request merged**
